src/services/redeem_and_unstake/ProgressService.ts

```diff
@@ -0,0 +1,158 @@
+import assert from 'assert';
+import Web3 from 'web3';
+
+import { interacts } from '@openst/mosaic-contracts';
+
+import { AUXILIARY_GAS_PRICE, ORIGIN_GAS_PRICE } from '../../Constants';
+import Logger from '../../Logger';
+import Gateway from '../../models/Gateway';
+import Message from '../../models/Message';
+import GatewayRepository from '../../repositories/GatewayRepository';
+import {MessageStatus, MessageType} from '../../repositories/MessageRepository';
+import Utils from '../../Utils';
+
+/**
+ * It facilitates progress redeeming and unstaking.
+ */
+export default class ProgressService {
+  private originWeb3: Web3;
+
+  private auxiliaryWeb3: Web3;
+
+  private coGatewayAddress: string;
+
+  private originWorkerAddress: string;
+
+  private auxWorkerAddress: string;
+
+  private gatewayRepository: GatewayRepository;
+
+  /**
+   * @param gatewayRepository Instance of GatewayRepository.
+   * @param originWeb3 Origin chain web3 object.
+   * @param auxiliaryWeb3 Auxiliary chain web3 object.
+   * @param coGatewayAddress Address of coGateway contract.
+   * @param originWorkerAddress Origin chain worker address.
+   * @param auxWorkerAddress Auxiliary chain worker address.
+   */
+  public constructor(
+    gatewayRepository: GatewayRepository,
+    originWeb3: Web3,
+    auxiliaryWeb3: Web3,
+    coGatewayAddress: string,
+    originWorkerAddress: string,
+    auxWorkerAddress: string,
+  ) {
+    this.gatewayRepository = gatewayRepository;
+    this.originWeb3 = originWeb3;
+    this.auxiliaryWeb3 = auxiliaryWeb3;
+    this.coGatewayAddress = coGatewayAddress;
+    this.originWorkerAddress = originWorkerAddress;
+    this.auxWorkerAddress = auxWorkerAddress;
+  }
+
+  /**
+   * This method react on changes in Message model.
+   * @param messages List of Message models.
+   */
+  public async update(messages: Message[]): Promise<void> {
+    Logger.debug('Progress service invoked');
+
+    const progressPromises = messages
+      .filter(message => message.type === MessageType.Redeem)
+      .filter(message => message.isValidSecret())
```

Comment:
both filters can be merged to save the extra loop.

Reply:
Done!